src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+// Implements checking before attribute writes.
+CHIP_ERROR Instance::Write(const ConcreteDataAttributePath & attributePath, AttributeValueDecoder & aDecoder)
+{
+    DataModel::Nullable<uint8_t> newMode;
+    ReturnErrorOnFailure(aDecoder.Decode(newMode));
+    Status status;
+
+    if (!newMode.IsNull()) // This indicates that the new mode is null.
+    {
+        if (!IsSupportedMode(newMode.Value()))
+        {
+            return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+        }
+    }
+
+    switch (attributePath.mAttributeId)
+    {
+    case ModeBase::Attributes::StartUpMode::Id:
+        status = UpdateStartUpMode(newMode);
+        return StatusIB(status).ToChipError();
+    case ModeBase::Attributes::OnMode::Id:
+        status = UpdateOnMode(newMode);
+        return StatusIB(status).ToChipError();
+    }
+
+    return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+}
+
+Status Instance::UpdateCurrentMode(uint8_t aNewMode)
+{
+    if (!IsSupportedMode(aNewMode))
+    {
+        return Protocols::InteractionModel::Status::InvalidValue;
+    }
+    uint8_t oldMode = mCurrentMode;
+    mCurrentMode    = aNewMode;
+    if (mCurrentMode != oldMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::CurrentMode::Id), mCurrentMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CurrentMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode)
+{
+    if (!aNewStartUpMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewStartUpMode.Value()))
+        {
+            return Protocols::InteractionModel::Status::InvalidValue;
+        }
+    }
+    DataModel::Nullable<uint8_t> oldStartUpMode = mStartUpMode;
+    mStartUpMode                                = aNewStartUpMode;
+    if (mStartUpMode != oldStartUpMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::StartUpMode::Id), mStartUpMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::StartUpMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateOnMode(DataModel::Nullable<uint8_t> aNewOnMode)
+{
+    if (!aNewOnMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewOnMode.Value()))
```

Comment:
The UpdateCurrentMode comments apply here.

Reply:
Done.